test/test_dynamic_shapes.py

```diff
@@ -678,6 +681,51 @@ def test_statically_known_true(self):
         # No guards should be generated
         self.assertEqual(len(shape_env.guards), 0)
 
+    def test_ephemeral_source_simplification(self):
+        from torch._dynamo.source import EphemeralSource
+
+        for x_first_in_check in [False, True]:
+            shape_env = ShapeEnv()
+            shape = (5, 10)
+            dynamic_dims = [DimDynamic.DYNAMIC for _ in shape]
+            x = create_symbolic_tensor(
+                "x",
+                torch.randn(*shape),
+                shape_env,
+                source=EphemeralSource(),
+                dynamic_dims=dynamic_dims,
+            )
+            y = create_symbolic_tensor(
+                "y",
+                torch.randn(*shape),
+                shape_env,
+                dynamic_dims=dynamic_dims,
+            )
+
+            def _get_ephemeral_source_symbols(t):
+                return [
+                    s.node._expr for s in (*t.shape, *t.stride(), t.storage_offset())
+                    if isinstance(s, torch.SymInt) and s.node._expr in shape_env.var_to_sources
+                    and any(
+                        source.is_ephemeral() for source in shape_env.var_to_sources[s.node._expr]
+                    )
+                ]
+
+            # these checks should simplify out the ephemeral symbols, regardless of the
```

Comment:
you should double check this test actually fails without your special sorter. I think it does, because y lexicographically sorts after x which means in reverse sort it would be preferentially picked first in all tie breaks... but if you had swapped the names then this test would not actually test what you want, as we would always consistently eliminate the ephemeral variable by accident (by virtue of its naming).

Actually, our heuristic where we try to substitute newer variables with older ones is pretty good, it probably is enough to eliminate ephemerals (except it doesn't actually work, because we string order rather than order on the allocation number LOL)